### ISSUE TITLE
max-statements-per-line: 2

### ISSUE THREAD

**Issue by christianbundy** [standard/standard#1431]
**What version of this package are you using?** 14.3.1

**What problem do you want to solve?** I'd like to limit the number of statements per line.

**What do you think is the correct solution to this problem?**

```json
{
  "max-statements-per-line": ["error", { "max": 2 }]
}
```

Previously a maximum of one statement per line was suggested as https://github.com/standard/standard/issues/798, but this is very aggressive. Instead, I'd like to propose a maximum of two statements per line. This should allow:

```javascript
// ✔ 2
if (err) { throw err }
// ✔ 2
list.map((item) => item.reverse())
```

But disallow:

```javascript
// ✘ 3
if (err) { throw err } else { return 42 }
// ✘ 5
function isJson (str) { try { JSON.parse(str) } catch (e) { return false } return true }
```

**Are you willing to submit a pull request to implement this change?**

Yes! 

---

Ecosystem impact:

```console
$ npm t
echint/lib/bin.js:14:111: This line has 3 statements. Maximum allowed is 2.
standard-engine/test/clone.js:58:41: This line has 3 statements. Maximum allowed is 2.
markdown-pdf/phantom/render.js:84:31: This line has 5 statements. Maximum allowed is 2.
markdown-pdf/test/index.js:94:78: This line has 5 statements. Maximum allowed is 2.
markdown-pdf/test/index.js:110:80: This line has 5 statements. Maximum allowed is 2.
aerospike/test/client.js:110:82: This line has 3 statements. Maximum allowed is 2.
levelup/test/benchmarks/index.js:59:94: This line has 3 statements. Maximum allowed is 2.
testdouble/test/safe/callback.test.js:65:61: This line has 3 statements. Maximum allowed is 2.
testdouble/test/safe/config.test.js:34:52: This line has 3 statements. Maximum allowed is 2.
testdouble/test/unit/when/chain-stubbing.test.js:13:38: This line has 3 statements. Maximum allowed is 2.

1..220
# tests 220
# pass  214
# fail  6
```

**Comment by mightyiam** [comment]
Thank you for the idea, @christianbundy.

I do use 2 statements per line quite often, such as `if (foo) return`. I do not remember when's the last time I have used 3 statements in one line.

What anti-pattern are we trying to prevent with this? Here are examples from the [rule docs](https://eslint.org/docs/rules/max-statements-per-line):

```js
/*eslint max-statements-per-line: ["error", { "max": 2 }]*/

var bar; var baz; var qux;
if (condition) { bar = 1; } else { baz = 2; }
for (var i = 0; i < length; ++i) { bar = 1; baz = 2; }
switch (discriminant) { case 'test': break; default: break; }
function foo() { bar = 1; baz = 2; }
var foo = function foo() { bar = 1; };
(function foo() { bar = 1; baz = 2; })();
```

I never do any of those.

**Comment by christianbundy** [comment]
Yep, exactly. I was looking at some non-StandardJS code and saw `if (a) { b()  } else { c() }` and realized that `eslint:recommended` reported this an an error. When I noticed that we didn't report this for this I figured I should open an issue, but saw the old issue and decided that we should probably allow two statements per line to keep the ecosystem impact small.

**Comment by mightyiam** [comment]
While I don't make final calls here, I assume that this would be accepted. Therefore if you'd like to contribute further, I suggest a pull request could be a good next step.

**Comment by feross** [comment]
Pull request is not necessary. I'll take a closer look at this when I have time to do another StandardJS release and I can PR it.

**Comment by christianbundy** [comment]
Thanks, please let me know if there's anything I can do to help out.

**Comment by feross** [comment]
In general, we try to avoid rules that have "magic numbers" in them since they may be hard to remember and we always want it to be possible to write code that passes `standard` without needing to rely on auto-fixing or a text editor with built-in support.

But I agree that some of the examples you posted are egregious. I just don't think we'd be able to enforce this without a lot of false positives – i.e. code that's actually easy-to-read but technically has 3+ statements.

@standard/team – any thoughts on this?

**Comment by mcollina** [comment]
I think the following is still readable:

`Object.keys(obj).map((a) => a.prop).reduce(fn)`

I feel 2 might be a bit too restrictive.

I think we can start adding it with a higher number (maybe 10?) we can agree to be always problematic. I've seen this abused by some teams.

**Comment by Raynos** [comment]
Consider the number you use for function parameters ( https://eslint.org/docs/rules/max-params ).

ESLint says 4 function parameters in a method or function is too much.

Making 4 statements too many seems reasonable.

I do notice that `standard` has no opinion on `max-params` so it might also make sense for it to have no opinion on `max-statements-per-line`.

**Comment by LinusU** [comment]
Personally leaning against this because of the "magic number" argument

**Comment by ungoldman** [comment]
> I do notice that standard has no opinion on max-params so it might also make sense for it to have no opinion on max-statements-per-line.

👍 feels like an overreach to me fwiw
